pkg/controller/persistentvolumelabel/pvlcontroller.go

```
@@ -0,0 +1,368 @@
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.ebsVolumes == nil {
+		awsCloudProvider, ok := pvlc.cloud.(*aws.Cloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving AWS cloud provider")
+		}
+		pvlc.ebsVolumes = awsCloudProvider
+	}
+	return pvlc.ebsVolumes, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findGCEPDLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.GCEPersistentDisk.PDName == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+
+	provider, err := pvlc.getGCECloudProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the zone is already labeled, honor the hint
+	zone := volume.Labels[kubeletapis.LabelZoneFailureDomain]
+
+	labels, err := provider.GetAutoLabelsForPD(volume.Spec.GCEPersistentDisk.PDName, zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getGCECloudProvider returns the GCE cloud provider, for use for querying volume labels
+func (pvlc *PersistentVolumeLabelController) getGCECloudProvider() (*gce.GCECloud, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.gceCloudProvider == nil {
+		gceCloudProvider, ok := pvlc.cloud.(*gce.GCECloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving GCE cloud provider")
+		}
+		pvlc.gceCloudProvider = gceCloudProvider
+	}
+	return pvlc.gceCloudProvider, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) updateVolume(volName string, volLabels map[string]string) error {
+	glog.V(4).Infof("updating PersistentVolume %s", volName)
+	curVol, err := pvlc.kubeClient.Core().PersistentVolumes().Get(volName, metav1.GetOptions{IncludeUninitialized: true})
+	if err != nil {
+		return err
+	}
+	objCopy := curVol.DeepCopyObject()
+	newVolume, ok := objCopy.(*v1.PersistentVolume)
+	if !ok {
+		return fmt.Errorf("failed to cast copy into persistentvolume object %#v", newVolume)
```

I would let it panic.  That would be a pretty severe programmer error.